java-checks/src/main/java/org/sonar/java/checks/StaticMemberAccessCheck.java

```diff
@@ -40,32 +41,49 @@ public class StaticMemberAccessCheck extends IssuableSubscriptionVisitor {
     .withAnyParameters()
     .build();
 
+  private String currentPackage = "";
+
   @Override
   public List<Tree.Kind> nodesToVisit() {
-    return Collections.singletonList(Tree.Kind.MEMBER_SELECT);
+    return List.of(Tree.Kind.MEMBER_SELECT, Tree.Kind.COMPILATION_UNIT);
   }
 
 
   @Override
   public void visitNode(Tree tree) {
-    MemberSelectExpressionTree mse = (MemberSelectExpressionTree) tree;
-    Symbol symbol = mse.identifier().symbol();
-    if (symbol.isStatic()  && !isListOrSetOf(mse)) {
-      ExpressionTree expression = mse.expression();
-      Type staticType = symbol.owner().type();
-      Type expressionType = expression.symbolType();
-      if (!staticType.isUnknown() && !expressionType.isUnknown()
-        && !expressionType.erasure().equals(staticType.erasure())) {
-        QuickFixHelper.newIssue(context)
-          .forRule(this)
-          .onTree(mse.identifier())
-          .withMessage("Use static access with \"%s\" for \"%s\".", staticType.fullyQualifiedName(), symbol.name())
-          .withQuickFix(() -> quickFix(expression, staticType))
-          .report();
+    if(tree instanceof CompilationUnitTree file){
+      PackageDeclarationTree packageDecl = file.packageDeclaration();
+      currentPackage = packageDecl==null ? "" : packageDecl.packageName().toString();
+    }else{
+      MemberSelectExpressionTree mse = (MemberSelectExpressionTree) tree;
+      Symbol symbol = mse.identifier().symbol();
+      if (symbol.isStatic()  && !isListOrSetOf(mse)) {
+        ExpressionTree expression = mse.expression();
+
+        Symbol owner = symbol.owner();
+        Type staticType = owner.type();
+        Type expressionType = expression.symbolType();
+
+        if (!staticType.isUnknown() && !expressionType.isUnknown()
+          && !expressionType.erasure().equals(staticType.erasure())
+          && (currentPackage.equals(classPackage(expressionType)) || owner.isPublic())
```

Comment:
Can we wrap this expression into a method with a meaningful name?

Reply:
:+1: